journal/src/main/scala/com/evolutiongaming/kafka/journal/util/ResourcePool.scala

```diff
@@ -0,0 +1,276 @@
+    }
+  }
+
+  case class State[F[_], O](
+    objects: Map[ObjectId, AllocatedObject[F, O]],
+    waiters: Map[WaiterId, Waiter[F, O]],
+  ) {
+
+    def findFreeObject: Option[(ObjectId, AllocatedObject[F, O])] =
+      objects.collectFirst {
+        case (id, obj) if obj.state.isFree => (id, obj)
+      }
+
+    def markObjectAsUsed(id: ObjectId, obj: AllocatedObject[F, O]): State[F, O] =
+      State(
+        objects + (id -> obj.copy(state = AllocatedObject.State.Used)),
+        waiters
+      )
+
+    def markObjectAsFree(id: ObjectId, obj: AllocatedObject[F, O], releasedAt: FiniteDuration): State[F, O] =
+      State(
+        objects + (id -> obj.copy(state = AllocatedObject.State.Free(releasedAt))),
+        waiters
+      )
+
+    def addObject(id: ObjectId, obj: AllocatedObject[F, O]): State[F, O] =
+      State(
+        objects + (id -> obj),
+        waiters
+      )
+
+    def removeObject(id: ObjectId): State[F, O] =
+      State(
+        objects - id,
+        waiters
+      )
+
+    def addWaiter(id: WaiterId, waiter: Waiter[F, O]): State[F, O] =
+      State(
+        objects,
+        waiters + (id -> waiter),
+      )
+
+    def removeWaiter(id: WaiterId): State[F, O] =
+      State(
+        objects,
+        waiters - id
+      )
+  }
+
+  def fixedSize[F[_]: Async, O](
+    poolSize: Int,
+    acquireTimeout: FiniteDuration,
+    idleTimeout: FiniteDuration,
+    log: Log[F],
+  )(
+    resource: Resource[F, O]
+  ): Resource[F, ResourcePool[F, O]] = {
+
+    type StateRef = Ref[F, State[F, O]]
+
+    def genUuid = Sync[F].delay(UUID.randomUUID())
+
+    def poolImpl(mutex: Mutex[F], stateRef: StateRef): ResourcePool[F, O] = {
+
+      val allocateObject: F[(ObjectId, AllocatedObject[F, O])] =
+        for {
+          id <- genUuid
+          a <- resource.allocated
+          (obj, release) = a
+        } yield id -> AllocatedObject(obj, release, AllocatedObject.State.Used)
+
+      def checkIfTimedOut(
+        waiterId: WaiterId,
+        waiter: Waiter[F, O]
+      ): F[Unit] = {
+        waiter.tryGet.flatMap {
+          case Some(_) =>
+            Async[F].unit
+          case None =>
+            mutex.lock.surround {
+              waiter.tryGet.flatMap {
+                case Some(_) =>
+                  Async[F].unit
+                case None =>
+                  for {
+                    _ <- waiter.complete(
+                      Left(
+                        new TimeoutException(s"Resource acquisition timed out")
+                      )
+                    )
+                    _ <- stateRef.update(_.removeWaiter(waiterId))
+                  } yield ()
+              }
+            }
+        }
+      }
+
+      val acquireWaiter: F[Waiter[F, O]] =
+        Deferred[F, Either[Throwable, (ObjectId, O)]].flatTap { waiter =>
+          mutex.lock.surround {
+            stateRef.get.flatMap { state =>
+              state.findFreeObject match {
```

Comment:
I propose that we check `state.objects.size < poolSize` and only there after try to `findFreeObject`